sssd_test_framework/utils/authentication.py

```diff
@@ -213,47 +232,138 @@ def password_expired(self, username: str, password: str, new_password: str) -> b
-                timeout {{puts "expect result: Unexpected output"; exit 1}}
-                eof {{puts "expect result: Unexpected end of file"; exit 2}}
+                timeout {{puts "expect result: Unexpected output"; exit 201}}
+                eof {{puts "expect result: Unexpected end of file"; exit 202}}
             }}
 
             expect {{
                 -re $prompt {{puts "expect result: Password change was successful"; exit 0}}
-                timeout {{puts "expect result: Unexpected output"; exit 1}}
-                eof {{puts "expect result: Unexpected end of file"; exit 2}}
+                timeout {{puts "expect result: Unexpected output"; exit 201}}
+                eof {{puts "expect result: Unexpected end of file"; exit 202}}
             }}
 
             puts "expect result: Unexpected code path"
-            exit 3
+            exit 203
+        """
+        )
+
+        if result.rc > 200:
+            raise ExpectScriptError(result.rc)
+
+        return result.rc == 0
+
+    def passkey(self, username: str, *, pin: str | int, device: str, ioctl: str, script: str) -> bool:
+        """
+        Call ``su - $username`` and authenticate the user with passkey.
+
+        :param username: User name
+        :type username: str
+        :param pin: Passkey PIN.
+        :type pin: str | int
+        :param device: Path to local umockdev device file.
+        :type device: str
+        :param ioctl: Path to local umockdev ioctl file.
+        :type ioctl: str
+        :param script: Path to local umockdev script file
+        :type script: str
+        :return: Generated passkey mapping string.
+        :rtype: str
+        :return: True if authentication was successful, False otherwise.
+        :rtype: bool
         """
+        self.fs.backup("/etc/sysconfig/sssd")
+        device_path = self.fs.upload_to_tmp(device, mode="a=r")
+        ioctl_path = self.fs.upload_to_tmp(ioctl, mode="a=r")
+        script_path = self.fs.upload_to_tmp(script, mode="a=r")
+
+        run_su = self.fs.mktmp(
+            rf"""
+            #!/bin/bash
+            set -ex
+            env | grep ^UMOCKDEV_ > /etc/sysconfig/sssd
+            printf "LD_PRELOAD=$LD_PRELOAD" >> /etc/sysconfig/sssd
+            systemctl restart sssd
+            chmod -R a+rwx $UMOCKDEV_DIR
+
+            su --shell /bin/sh nobody -c "su - '{username}'"
+            """,
+            mode="a=rx",
+        )
+
+        playback_umockdev = self.fs.mktmp(
+            rf"""
+            #!/bin/bash
+
+            LD_PRELOAD=/opt/random.so umockdev-run \
+                --device '{device_path}' \
+                --ioctl '/dev/hidraw1={ioctl_path}' \
+                --script '/dev/hidraw1={script_path}' \
+                -- '{run_su}'
+            """,
+            mode="a=rx",
+        )
+
+        result = self.host.ssh.expect(
+            rf"""
+            # It takes some time to get authentication failure
+            set timeout 10
```

Comment:
Could we use a default argument for the timeout here?

Reply:
Done.